libs/src/services/contracts/ProviderSelection.js

```diff
@@ -0,0 +1,79 @@
+
+const Web3 = require('web3')
+
+const ServiceSelection = require('../../service-selection/ServiceSelection')
+const EthWeb3Manager = require('../../../src/services/ethWeb3Manager/index')
+
+/**
+ * This class provides the logic in selecting the proper gateway
+ */
+// TODO: use env vars idk how to use env vars lol
+const gatewayProvidersList = ['https://poa-gateway.audius.co', 'https://core.poa.network']
+
+class ProviderSelection extends ServiceSelection {
+  constructor (whitelist = gatewayProvidersList) {
+    super({
+      whitelist,
+      // TODO: return env vars
+      getServices: async () => { return whitelist }
+    })
+  }
+
+  /**
+   * Initializes the ContractClient with a healthy POA provider.
+   *
+   * First, try the contract logic with currently set provider when web3Manager
+   * was initialized. If contract logic fails, add current provider to unhealthy
+   * list and retry with the other provided gateways.
+   *
+   * If all gateways have been tried contract logic fails (gateways are unhealthy),
+   * log an error.
+   * @param {*} contractClient object used for making transaction calls
+   */
+  async setContractClientProvider (contractClient) {
+    contractClient.setIsInitializing(true)
+
+    try {
+      const contractAddress = await contractClient.getRegistryAddress(contractClient.contractRegistryKey)
```

Comment:
I *think* we should wrap all this code in something inside contract client rather than having this class do all this logic?

Reply:
yeah that's a good point. we can leave the actual contract logic back in the contract client class. and only have the providerselection class do the provider selcting